File: Leetcode/897_Increasing_Order_Search_Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var increasingBST = function (root) {
    const result = [];
    traversal(root, result);
    let newRoot = null;
    let beforeNode = null;
    result.forEach(val => {
        if (newRoot === null) {
            newRoot = new TreeNode(val);
            newRoot.right = null;
            newRoot.left = null;
            beforeNode = newRoot;
        } else {
            beforeNode.right = new TreeNode(val);
            beforeNode.left = null;
            beforeNode = beforeNode.right;
        }
    })

    return newRoot;
};

const traversal = (root, result) => {
    if (root) {
        traversal(root.left, result);
        result.push(root.val);
        traversal(root.right, result);
    }
}

File: Leetcode/0897/897_Increasing_Order_Search_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* pre;
    TreeNode* start;
    TreeNode* increasingBST(TreeNode* root) {
        if(root == nullptr)
        {
            return nullptr;
        }
       
        increasingBST(root->left);
        if(start == nullptr)
        {
            TreeNode* t = new TreeNode(root->val);
            start = t;
            pre = start;
        }
        else
        {
            TreeNode* t = new TreeNode(root->val);
            pre->right = t;
            pre = t;
        }
        increasingBST(root->right);
        
        return start;
    }
};